### PR TITLE
extractorapp: Fix reprojection issues with WFS extractions

### LINKED ISSUE #1906: Extractor : wrong SRS for vector data
I tried to extract data from a GeoServer layer coming from the dev server of GéoPicardie.
This layer is stored as a shapefile in EPSG:2154 projection.
I chose to extract it as a shapefile in WGS84 SRS.
The .proj files of the data layer and the bbox are coherent with WGS84.
The data layer coordinates seem to be totally incoherent with this projection. See the ogrinfo command result obtained with the extracted data:

```
Layer name: olv_equ_bdolv_geo_smvo_dechetterie_POINT
{{ Geometry: 3D Multi Point}}
{{ Feature Count: 24}}
{{ Extent: (626997.320938, 6889290.994452) - (704022.321241, 6951376.923891)}}
{{ Layer SRS WKT:}}
{{ GEOGCS["WGS 84",}}
{{ DATUM["World Geodetic System 1984",}}
{{ SPHEROID["WGS 84",6378137.0,298.257223563,}}
{{ AUTHORITY["EPSG","7030"]],}}
{{ AUTHORITY["EPSG","6326"]],}}
{{ PRIMEM["Greenwich",0.0,}}
{{ AUTHORITY["EPSG","8901"]],}}
{{ UNIT["degree",0.017453292519943295],}}
{{ AXIS["Geodetic longitude",EAST],}}
{{ AXIS["Geodetic latitude",NORTH],}}
{{ AUTHORITY["EPSG","4326"]]}}
```

[for Camptocamp developers, the original description of the issue is there (in french): https://jira.camptocamp.com/browse/GSPIC17-129]

### PR DESCRIPTION
**Pull request by groldan**
Fixes #1906 
The WFS DataStore on the GeoTools version used (9.2) does not handle reprojection, so it's done in-process

### CODE REVIEW

**Comment by groldan** [comment]
Marked as WIP, tested locally but couldn't run docker-compose. Getting ```javax.net.ssl.SSLHandshakeException: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target```

@fvanderbiest do you know if I need to change some configuration or anything?

**Comment by fvanderbiest** [comment]
Thanks Gabriel !
Can you rebase on top of `19.04` ?

SSLHandshakeException probably comes from a mistake in `casTicketValidationUrl` in the datadir. Please revert my recent change for the docker use-case and check this works for you.

**Comment by groldan** [comment]
@fvanderbiest rebased on top of 19.04. Shall the PR target the 19.04 branch as well?

**Comment by fvanderbiest** [comment]
> Shall the PR target the 19.04 branch as well?

Yes, so we have this important fix in the upcoming release.
Thanks !

File: extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java [hunk 1]
```diff
@@ -308,47 +326,52 @@ public void exceptionOccurred (Throwable exception) {
...
+		// current version (9.2) of WFS datastore does not perform reprojection
+		if (!CRS.equalsIgnoreMetadata(targetCrs, returnedCrs)) {
+			features = new ReprojectingFeatureCollection(features, targetCrs);
+		}
+		
+		return features;
+	}
+
 	/* This method is default for testing purposes */
     Query createQuery (ExtractorLayerRequest request, FeatureType schema) throws IOException, TransformException,
```
Comment (cmangeat) [inline extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java]: I didn't find the "createQueryTest", may be private ?

Seems that there is no more IOException anymore.

File: extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java [hunk 2]
```diff
@@ -308,47 +326,52 @@ public void exceptionOccurred (Throwable exception) {
...
-            String[] propArray = properties.toArray (new String[properties.size ()]);
-            Query query = new Query (request.getWFSName(), filter, propArray);
-
-            query.setCoordinateSystemReproject (request._projection);
-
-            return query;
-        default:
-            return null;
-        }
+		final CoordinateReferenceSystem nativeCrs = schema.getCoordinateReferenceSystem();
```
Comment (cmangeat) [inline extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java]: nativeCrs should be declared in the else part of the indirection

File: extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java [hunk 3]
```diff
@@ -213,9 +228,13 @@ public static void enablePreemptiveBasicAuth(URL capabilitiesURL, HttpClientBuil
      * @return the directory that contains the extracted file
      */
     public File extract (ExtractorLayerRequest request) throws IOException, TransformException, FactoryException {
-        if (request._owsType != OWSType.WFS) {
+    	checkNotNull(request);
+    	if (request._owsType != OWSType.WFS) {
```
Comment (cmangeat) [inline extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java]: https://github.com/georchestra/georchestra/pull/2582/commits/94bc8f06d32aeb61271f42d3dd01cf41bd9baec0#diff-812f3e1b48408b609afa21a720287c4aR228 tended to unify contract reading.

File: extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java [hunk 4]
```diff
@@ -80,6 +80,8 @@
 
     protected static final Log LOG = LogFactory.getLog(WcsExtractor.class.getPackage().getName());
 
+    private static final FilterFactory2 filterFactory = CommonFactoryFinder.getFilterFactory2 (GeoTools.getDefaultHints ());
```
[the remark this answers is not in the document]
Comment (groldan) [reply extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java]: fixed

File: extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java [hunk 5]
```diff
@@ -81,6 +84,7 @@
     protected static final Log LOG = LogFactory.getLog(WcsExtractor.class.getPackage().getName());
 
     private static final FilterFactory2 filterFactory = CommonFactoryFinder.getFilterFactory2 (GeoTools.getDefaultHints ());
+    private static final Set<String> supportedFormats = ImmutableSet.of("shp", "mif", "tab", "kml");
```
[the remark this answers is not in the document]
Comment (groldan) [reply extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java]: fixed

File: extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java [hunk 6]
```diff
@@ -218,6 +222,8 @@ public File extract (ExtractorLayerRequest request) throws IOException, Transfor
         if (request._owsType != OWSType.WFS) {
             throw new IllegalArgumentException (request._owsType + "must be WFS for the WfsExtractor");
         }
+		Preconditions.checkArgument(request._format != null && supportedFormats.contains(request._format.toLowerCase()),
+				"%s is not a recognized vector format", request._format);
```
[the remark this answers is not in the document]
Comment (groldan) [reply extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java]: fixed

File: extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java [hunk 7]
```diff
@@ -296,11 +302,10 @@ public void exceptionOccurred (Throwable exception) {
         } else if ("tab".equalsIgnoreCase(request._format)) {
             featuresWriter = new OGRFeatureWriter(progressListener, sourceSchema,  basedir, OGRFeatureWriter.FileFormat.tab, features);
             bboxWriter = new BBoxWriter(request._bbox, basedir, OGRFeatureWriter.FileFormat.tab, request._projection, progressListener );
-        } else if ("kml".equalsIgnoreCase(request._format)) {
+        } else {
+        	Preconditions.checkState("kml".equalsIgnoreCase(request._format));
```
[the remark this answers is not in the document]
Comment (groldan) [reply extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java]: fixed

File: extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java [hunk 8]
```diff
@@ -224,6 +225,7 @@ public File extract (ExtractorLayerRequest request) throws IOException, Transfor
         }
 		Preconditions.checkArgument(request._format != null && supportedFormats.contains(request._format.toLowerCase()),
 				"%s is not a recognized vector format", request._format);
+    	Preconditions.checkArgument(OWSType.WFS.equals(request._owsType));
```
[the remark this answers is not in the document]
Comment (groldan) [reply extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java]: outdated, precondition checked before

File: extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java [hunk 9]
```diff
@@ -45,6 +45,7 @@
 import org.geotools.geometry.jts.ReferencedEnvelope;
 import org.geotools.referencing.CRS;
 import org.geotools.util.NullProgressListener;
+import org.geotools.wfs.v2_0.WFS;
```
[the remark this answers is not in the document]
Comment (groldan) [reply extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java]: fixed imports

File: extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java (hunk 1 above)
Comment (groldan) [reply extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java]: I didn't find it either, the method was already there. We should certainly add one though

Comment (groldan) [reply extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java]: made it private and removed unnecessary IOException from throws list

File: extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java (hunk 2 above)
Comment (groldan) [reply extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java]: done

File: extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java (hunk 3 above)
Comment (groldan) [reply extractorapp/src/main/java/org/georchestra/extractorapp/ws/extractor/WfsExtractor.java]: I'm not sure what the problem is here?

**Pull request merged**